Rearrange / rename API reference docs

HalfWhitt opened the pull request:
Fixes #3941

## Changes

- Rearranges the API reference docs as listed in https://github.com/beeware/toga/issues/3941#issuecomment-3688146180
- Also moves the Architecture page from /reference/internals/architecture to /topics/architecture
- Adds redirects for all moves, using [mkdocs-redirects](https://github.com/mkdocs/mkdocs-redirects)
- Excludes snippets from sitemap, using [mkdocs-exclude](https://github.com/apenwarr/mkdocs-exclude)
  - Unfortunately there doesn't seem to be a way of excluding the nav file, SUMMARY.md, without breaking literate-nav and getting just the default auto-generated navigation. It's not a file *meant* for public consumption, but I don't suppose it does any harm for it to be listed...

I've checked the redirects by downloading the 0.5.3 sitemap and using the following script:
```python
import xml.etree.ElementTree as ET

import httpx
from rich import print

tree = ET.parse("053 sitemap.xml")
root = tree.getroot()

for child in root:
 	path = child[0].text.removeprefix("https://toga.beeware.org/")
 	status = httpx.head(f"http://127.0.0.1:8000/{path}").status_code
 	color = "green" if status == 200 else "red"
 	print(f"{path}: [{color}]{status}[/{color}]")
```
I also turned on traffic analytics on readthedocs, so I could [see what's been getting 404 errors](https://app.readthedocs.org/dashboard/toga/traffic-analytics/) and add those too.

## Remaining questions

- Whether Document goes in Resources or Application Components.
- Finalizing the shortened API descriptions, and/or keeping both a long and a short version.
  
   Now that I've tried it out, I think sticking to one consistent (brief) description is still the way to go; it reads more like the top line of a docstring. This may necessitate reintroducing some of the removed detail at the beginning of the Usage section. I've waited to do this, or to spend much time going over and finalizing shortened versions, before getting feedback on whether the general approach looks good to others.
- How (or if) to visually make the Container Widgets table be visually a subcategory of Widgets.

  I've tried using a lower-level heading:

  <img width="883" height="1112" alt="subheading" src="https://github.com/user-attachments/assets/938336d3-c913-41e3-9c97-0c797fbc9402" />

  And even mucking about with a faux-header insert:

  <img width="891" height="939" alt="inserted header" src="https://github.com/user-attachments/assets/73e12120-ec47-474c-a9cb-8fff0cf06637" />

  I don't think either of these look good or read as intended, and the latter exacerbates how tall the Widgets table already is, requiring even more scrolling up to find the header labels. (Unfortunately, sticky headers [doesn't seem to be possible](https://github.com/squidfunk/mkdocs-material/issues/4347) with mkdocs Materials.)

  It's already structurally a subcategory, and reads as such in the sidebar nav. Perhaps that's enough?

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

kattni commented:
I will take a look this evening.

kattni reviewed (changes requested):
I somehow failed to click "Finish" in this. I didn't notice until I noticed that my email didn't have a new reply for this PR. 🤦🏻‍♀️ 

I have one suggestion below. Otherwise, we discussed everything else.

HalfWhitt commented:
Another day, another merge conflict. This should be ready again.

freakboy3742 commented on docs/en/reference/api/data-representation/sources/listsource.md:
```diff
@@ -1,8 +1,10 @@
-{{ component_header("ListSource") }}
+# ListSource
+
+foo
```
bar? :-)

freakboy3742 commented on docs/en/reference/api/data-representation/sources/treesource.md:
```diff
@@ -1,8 +1,10 @@
-{{ component_header("TreeSource") }}
+# TreeSource
+
+foo
```
bar? :-)

HalfWhitt commented:
> - I think the extra depth for Source obscured more than it helped.

Fair enough!

> - I've also simplified "Container Widgets" to "Containers" - to me, Box et al are either widgets, or they're containers; "Container Widgets" seems like hedging our bets.

I still [think it's important](https://github.com/beeware/toga/issues/3941#issuecomment-3670965738) to point out that they are, in fact, widgets (and Kattni [agrees](https://github.com/beeware/toga/issues/3941#issuecomment-3677092226) too). If they're not going to be a subgroup of Widgets, the title seems like the only good way to call that out.

[Edit: For some reason those links are just going to the issue when I click on them normally, and only anchoring to the relevant comment if I open them in a new tab. No idea what's going on, or if that's just me somehow.]

> My immediate thought was that both of these should be subheadings of Data Rep and Widgets respectively (and that's what the most recent conversation on #3941 suggested); I'm not sure if the flat structure in the PR was intentional or accidental, but I experimented a bit with a deeper structure and I don't think having depth actually helped.

I tried a couple of ways of showing that containers were a subgroup of widgets while still displaying them on the index page (screenshotted in the initial post), and they looked weird. But current you agrees with past you — in our video call you recommended leaving them separate. ; )

> At this point, I'm happy; I won't merge so that @mhsmith has a chance to confirm he's happy with the details of the new structure.

👍 

> bar? :-)

Drat! Good catch. When I removed them from the YAML, I forgot to finish inlining their descriptions. But I see that's a moot point now anyway.

> `+ # Data Sources { #topic-data-sources }`

Okay, I see that I didn't update a couple of the links. But why go back to linking to an h1 anchor instead of to the page itself?

freakboy3742 commented:
> On a related note: `docs-lint` doesn't fail when internal links (either to pages or cross-references) can't be resolved. We should fix that, if possible.

No... but from Mkdocs perspective, that's a build issue, not a linting issue. AFAICT, a `docs-live` build *warns* about dead internal links; a `docs-all` build raises an error. Or have you found a case where that isn't happening?

HalfWhitt commented:
> > On a related note: `docs-lint` doesn't fail when internal links (either to pages or cross-references) can't be resolved. We should fix that, if possible.
> 
> No... but from Mkdocs perspective, that's a build issue, not a linting issue. AFAICT, a `docs-live` build _warns_ about dead internal links; a `docs-all` build raises an error. Or have you found a case where that isn't happening?

Oh never mind, I see what happened. The latest build I pushed before your review passed checks and built a preview, even though I'd messed up links... but while the links weren't pointing to the data source topic page, they *were* instead resolving to the `Source` reference section. So as far as MkDocs knew, everything was fine.
